// cdk8s-cli/lib/config.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.addImportToConfig = exports.readConfigSync = exports.HelmChartApiVersion = exports.SynthesisFormat = void 0;
const fs = __importStar(require("fs-extra"));
const yaml = __importStar(require("yaml"));
const CONFIG_FILE = 'cdk8s.yaml';
var SynthesisFormat;
(function (SynthesisFormat) {
    SynthesisFormat["PLAIN"] = "plain";
    SynthesisFormat["HELM"] = "helm";
})(SynthesisFormat = exports.SynthesisFormat || (exports.SynthesisFormat = {}));
var HelmChartApiVersion;
(function (HelmChartApiVersion) {
    HelmChartApiVersion["V1"] = "v1";
    HelmChartApiVersion["V2"] = "v2";
})(HelmChartApiVersion = exports.HelmChartApiVersion || (exports.HelmChartApiVersion = {}));
function readConfigSync() {
    if (fs.existsSync(CONFIG_FILE)) {
        const config = {
            ...yaml.parse(fs.readFileSync(CONFIG_FILE, 'utf-8')),
        };
        return config;
    }
    return undefined;
}
exports.readConfigSync = readConfigSync;
async function addImportToConfig(source) {
    var _a, _b;
    if (!fs.existsSync(CONFIG_FILE)) {
        // cdk8s import might be executed outside of an application (i.e without a config file)
        return;
    }
    let curConfig = yaml.parse(fs.readFileSync(CONFIG_FILE, 'utf-8'));
    const curImports = (_a = curConfig.imports) !== null && _a !== void 0 ? _a : [];
    if (!curImports.includes(source)) {
        const importsList = (_b = curConfig.imports) !== null && _b !== void 0 ? _b : [];
        importsList.push(source);
        let config = {
            ...curConfig,
            imports: importsList,
        };
        await fs.outputFile(CONFIG_FILE, yaml.stringify(config));
    }
}
exports.addImportToConfig = addImportToConfig;
